all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,454 @@
+# External Client Configuration
+
+Notes for this proposal:
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+## Overview
+
+There are many ways to access Temporal programmatically including CLI, SDK, and other components that leverage those
+such as samples and Terraform providers.
+
+Temporal needs a way for users to provide client configuration without hardcoding in CLI arguments or SDK options. This
+will allow code to seamlessly switch between profiles such as between a dev-server and cloud. The same configuration
+must work across all SDKs/CLIs.
+
+### Why?
+
+What are the reasons for building this?
+
+* Users want to switch environments without changing their code, but today they cannot.
+* Our samples should be able to be run on multiple environments including cloud, but today they are localhost only.
```

Comment:
I suspect because it's an easier transition for them. If we have an environment we promote for easy use, it seems reasonable to make samples run there easily. Also as I understand it, for many "digital natives" (or saas-first or cloud-first or whatever you'd like to call them), they'd go straight to cloud without a local step as they do with many other products they use. Also, running a sample against cloud allows them to confirm they have their cloud namespace setup properly.

Reply:
Agree with @cretz.

- We may want to make it possible for someone to check out the repo in a VSCode cloud workspace and run samples. (E.g. because they like to work that way, or they are working on a non-conventional development environment such as a tablet or phone).
- We may want to create a playground app where users can run and edit our samples.
- An organization may be testing out their cloud namespace and want to use samples for that.
- An organization may be onboarding Temporal users to cloud namespaces and want them to try running samples against their cloud namespace.

Reply:
Thanks, I definitely buy some of these.  (Though, others are written from our perspective rather than the user's perspective)  The meta-point I want to make is that we should incorporate some of these user needs into our goals to make sure we understand priority and are solving their complete needs.  (i.e. Let's not just discuss it in a thread but build it into the doc.)

Here are some possible stories in addition to the ones y'all provided.

* A user is figuring out how to use Temporal copies code from these samples which uses these client configs into their production code and expects it to work well.
* A user has a multi-region setup and wants to vary some of their client connection config depending on the region.
* User is migrating from self-hosted to cloud and wants to dynamically flip between the two environments, e.g. using a feature flag.

We should err on the side of writing more stories down and then, if we don't want to cater to them, saying so explicitly.  This will make sure we're aligned.